### PR TITLE
Stub FUSE filesystem implementation

### PR DESCRIPTION
**Pull request by ridwanmsharif**
This change adds some of the foundation of the fuse filesystem.
The filesystem can now be mounted as well.

### CODE REVIEW

**Comment by avagin** [comment]
Cc: @jinmouil

File: pkg/sentry/fsimpl/fuse/fusefs.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	}
+	delete(mopts, "fd")
+
+	deviceDescriptor, err := strconv.ParseInt(deviceDescriptorStr, 10 /* base */, 32 /* bitSize */)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	taskCtxKey := kernel.CtxTask
+	taskCtxValue := ctx.Value(taskCtxKey)
```
Comment (avagin) [inline pkg/sentry/fsimpl/fuse/fusefs.go:95]: can we use kernel.TaskFromContext here?

File: pkg/sentry/syscalls/linux/vfs2/mount.go [hunk 2]
```diff
@@ -94,6 +93,12 @@ func Mount(t *kernel.Task, args arch.SyscallArguments) (uintptr, *kernel.Syscall
 	if flags&linux.MS_NOEXEC == linux.MS_NOEXEC {
 		opts.Flags.NoExec = true
 	}
+	if flags&linux.MS_NODEV == linux.MS_NODEV {
```
Comment (avagin) [inline pkg/sentry/syscalls/linux/vfs2/mount.go:96]: I think it is better to do this change on a separate PR.

File: pkg/sentry/fsimpl/fuse/fusefs.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+		fsopts.groupID = uint32(groupID)
+	}
+
+	rootMode := linux.FileMode(0777)
+	modeStr, ok := mopts["rootmode"]
+	if ok {
+		delete(mopts, "rootmode")
+		mode, err := strconv.ParseUint(modeStr, 8, 32)
+		if err != nil {
+			ctx.Warningf("fusefs.FilesystemType.GetFilesystem: invalid mode: %q", modeStr)
```
Comment (avagin) [inline pkg/sentry/fsimpl/fuse/fusefs.go:131]: in all other cases, you use log.Warningf but here ctx.Warningf. Why?
And I think you may want to add fsType.Warningf to avoid repeating the log message prefix.

File: pkg/sentry/syscalls/linux/vfs2/mount.go (hunk 2 above)
Comment (ridwanmsharif) [reply pkg/sentry/syscalls/linux/vfs2/mount.go:96]: We had to remove them from the `unsupportedFlags`  above as `libfuse` uses them to mount FUSE filesystems. Added a `TODO` to add support for it for other filesystems.

File: pkg/sentry/fsimpl/fuse/fusefs.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		delete(mopts, "group_id")
+		groupID, err := strconv.ParseUint(groupIDStr, 10, 32)
+		if err != nil {
+			log.Warningf("%s.GetFilesystem: invalid group_id: group_id=%s", fsType.Name(), groupIDStr)
+			return nil, nil, syserror.EINVAL
+		}
+		fsopts.groupID = uint32(groupID)
+	}
+
+	rootMode := linux.FileMode(0777)
```
Comment (craig08) [inline pkg/sentry/fsimpl/fuse/fusefs.go:124]: Are we default to set rootMode `0777`? `libfuse` seems to use the original mode of the mountpoint as its root mode. https://github.com/libfuse/libfuse/blob/master/util/fusermount.c#L1133

File: pkg/sentry/fsimpl/fuse/fusefs.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	modeStr, ok := mopts["rootmode"]
+	if ok {
+		delete(mopts, "rootmode")
+		mode, err := strconv.ParseUint(modeStr, 8, 32)
+		if err != nil {
+			log.Warningf("%s.GetFilesystem: invalid mode: %q", fsType.Name(), modeStr)
+			return nil, nil, syserror.EINVAL
+		}
+		rootMode = linux.FileMode(mode & 07777)
```
Comment (craig08) [inline pkg/sentry/fsimpl/fuse/fusefs.go:133]: `&^linux.S_IFMT`

File: pkg/sentry/fsimpl/fuse/fusefs.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+		opts:     fsopts,
+	}
+
+	fs.VFSFilesystem().Init(vfsObj, &fsType, fs)
+
+	// TODO: dispatch a FUSE_INIT request to the FUSE daemon server before
+	//  returning. Mount will not block on this dispatched request.
+
+	// root is the fusefs root directory.
+	defaultFusefsDirMode := linux.FileMode(0755)
```
Comment (craig08) [inline pkg/sentry/fsimpl/fuse/fusefs.go:160]: Is this `rootMode`?
